File: commons-aws/src/test/java/test/pl/wrzasq/commons/aws/MessageDispatcherTest.java
/*
 * This file is part of the pl.wrzasq.commons.
 *
 * @license http://mit-license.org/ The MIT license
 * @copyright 2020 © by Rafał Wrzeszcz - Wrzasq.pl.
 */

package test.pl.wrzasq.commons.aws;

import java.io.IOException;
import java.util.function.Function;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.Mockito;
import org.mockito.junit.jupiter.MockitoExtension;
import pl.wrzasq.commons.aws.MessageDispatcher;

@ExtendWith(MockitoExtension.class)
public class MessageDispatcherTest {
    @Mock
    private Function<String, String> messageHandler;

    @Mock
    private ObjectMapper objectMapper;

    @Test
    public void handle() throws JsonProcessingException {
        var expected = "test";

        var client = new MessageDispatcher<>(
            this.objectMapper,
            this.messageHandler
        );

        var message = 44;
        var serialized = "44";

        Mockito.when(this.objectMapper.writeValueAsString(message)).thenReturn(serialized);
        Mockito.when(this.messageHandler.apply(serialized)).thenReturn(expected);

        var response = client.send(message);

        Mockito.verify(this.messageHandler).apply(serialized);

        Assertions.assertEquals(
            expected,
            response,
            "MessageDispatcher.send() should return operation result."
        );
    }

    @Test
    public void handleInvalidJson() throws JsonProcessingException {
        var client = new MessageDispatcher<>(
            this.objectMapper,
            this.messageHandler
        );

        var message = 44;

        Mockito.when(this.objectMapper.writeValueAsString(message)).thenThrow(JsonProcessingException.class);

        Assertions.assertThrows(IllegalArgumentException.class, () -> client.send(message));
    }
}

File: commons-aws/src/test/kotlin/test/pl/wrzasq/commons/aws/messaging/MessageDispatcherTest.kt
/*
 * This file is part of the pl.wrzasq.commons.
 *
 * @license http://mit-license.org/ The MIT license
 * @copyright 2020 - 2022 © by Rafał Wrzeszcz - Wrzasq.pl.
 */

package test.pl.wrzasq.commons.aws.messaging

import io.mockk.every
import io.mockk.impl.annotations.MockK
import io.mockk.junit5.MockKExtension
import io.mockk.verify
import kotlinx.serialization.ExperimentalSerializationApi
import kotlinx.serialization.KSerializer
import kotlinx.serialization.SerializationException
import kotlinx.serialization.json.Json
import kotlinx.serialization.modules.SerializersModule
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows
import org.junit.jupiter.api.extension.ExtendWith
import pl.wrzasq.commons.aws.messaging.MessageDispatcher

@ExperimentalSerializationApi
@ExtendWith(MockKExtension::class)
class MessageDispatcherTest {
    @MockK
    lateinit var messageHandler: (String) -> String

    @MockK
    lateinit var serializerModule: SerializersModule

    @MockK
    lateinit var serializer: KSerializer<Any>

    @MockK
    lateinit var json: Json

    @Test
    fun handle() {
        val expected = "test"
        val client = MessageDispatcher(json, messageHandler)
        val message = 44
        val serialized = "44"

        every { json.serializersModule } returns serializerModule
        every { serializerModule.getContextual(Any::class, any()) } returns serializer
        every { json.encodeToString(serializer, message) } returns serialized
        every { messageHandler(serialized) } returns expected

        val response = client.send(message)

        verify { messageHandler(serialized) }

        assertEquals(expected, response)
    }

    @Test
    fun handleInvalidJson() {
        val client = MessageDispatcher(json, messageHandler)
        val message = 44

        every { json.serializersModule } returns serializerModule
        every { serializerModule.getContextual(Any::class, any()) } returns serializer
        every { json.encodeToString(serializer, message) } throws SerializationException()

        assertThrows<IllegalArgumentException> { client.send(message) }
    }
}
